sklearn/impute.py

```diff
@@ -1144,26 +1146,37 @@ def _get_missing_features_info(self, X):
             imputer_mask = sparse_constructor(
                 (mask, X.indices.copy(), X.indptr.copy()),
                 shape=X.shape, dtype=bool)
+            imputer_mask.eliminate_zeros()
 
-            missing_values_mask = imputer_mask.copy()
-            missing_values_mask.eliminate_zeros()
-            features_with_missing = (
-                np.flatnonzero(np.diff(missing_values_mask.indptr))
-                if missing_values_mask.format == 'csc'
-                else np.unique(missing_values_mask.indices))
+            if self.features in ('missing-only', 'not-constant'):
+                if imputer_mask.format == 'csc':
+                    n_missing = np.diff(imputer_mask.indptr)
+                else:
+                    n_missing = np.bincount(imputer_mask.indices,
+                                            minlength=X.shape[1])
 
             if self.sparse is False:
                 imputer_mask = imputer_mask.toarray()
             elif imputer_mask.format == 'csr':
                 imputer_mask = imputer_mask.tocsc()
         else:
             imputer_mask = _get_mask(X, self.missing_values)
-            features_with_missing = np.flatnonzero(imputer_mask.sum(axis=0))
+
+            if self.features in ('missing-only', 'not-constant'):
+                n_missing = imputer_mask.sum(axis=0)
 
             if self.sparse is True:
                 imputer_mask = sparse.csc_matrix(imputer_mask)
 
-        return imputer_mask, features_with_missing
+        if self.features == 'all':
+            features_indices = np.arange(X.shape[1])
+        else:
```

Comment:
this would be clearer as:
```
elif self.features == 'missing-only':
    features_indices = np.flatnonzero(n_missing)
else:
    features_indices = np.flatnonzero(np.logical_and(n_missing < X.shape[0], n_missing > 0))
```